_posts/2017-04-04-what-every-developer-should-know-about-ir.md

```diff
@@ -0,0 +1,250 @@
+
+| term    |  found at  |
+|---------|------------|
+| a       | 1,2        |
+| dog     | 1          |
+| eats    | 1,2        |
+| dog     | 1          |
+| food    | 1          |
+| cat     | 2          |
+| mice    | 2          |
+
+Our query matches to document 1, but not document 2, because eventhough "eats" is found in both documents "dog" exists only in document 1. Note that you could implement this *index* by yourself to database, but you're probably better of by using something from the box. Most DBMS:s have full text indexing capabilities out of the box, or you can use something like Elasticsearch, Lucene or SOLR. In this example, we wrote "a" to our index as well, but it is usually beneficial to include stop word list, which are excluded from the index and queries. Stop words are examined later more thoroughly.
+
+If we had bigger *collection*, could it be, that there exists relevant documents which do not talk about dogs with the word "dog", but refer to the same animal as "hound" or "mongrel" or "pooch" or even "stray". We're in luck! With boolean model we can craft a query combining synonyms within OR-facets such as:
+
+```
+(dog OR hound OR mongrel OR pooch OR stray)
+AND
+(diet OR nourishment OR nutrition OR sustenance)
+```
+
+Could some of the documents talk about *hounds* or *pooches* (in plural form) and not mention singular form at all? It's possible, hence:
+
+```
+(dog OR dogs OR hound OR hounds OR mongrel OR mongrels
+  OR pooch OR pooches OR stray)
+AND
+(diet OR nourishment OR nutrition OR sustenance)
+```
+
+Or we might even utilize wildcard operator:
+
+```
+(dog* OR hound* OR mongrel* OR pooch* OR stray*)
+AND
+(diet* OR nourishment* OR nutrition* OR sustenance*)
+```
+
+We begin to see, that we should be really good with words or even have a dictionary with us when we craft our queries. Further on we realise, that making queries like this, is a job for a professional information retrieval specialist. Building those OR-facets seems a horrible task and practically no system is anymore implemented like this. Yet, some of the largest and dominant systems still use boolean model as a main interface - it's fair to say, that these types of systems are meant for expert usage.
+
+There has been loads of systems that use synonym dictionaries to add synonyms automatically to queries, so user don't have to think the synonyms at all. This is called *query expansion*, which is very important technique when helping user with queries. The problem is that you need a really good synonym dataset to make something like this automatically. This is usually really beneficial in large collections that deal with wide range of subjects - such as news collections - and is not that beneficial in smaller rather heavily specialised collections where users usually know the terms documents contain. Query expansion is not limited to boolean model, but usable in various other retrieval models as well, which we'll take a peek next.
```

Comment:
so user doesn't have to think
which is _a_ very important technique when helping _the_ user with queries.
is not limited to _the_ boolean model